packages/driver/src/util/serialization/README.md

```diff
@@ -0,0 +1,96 @@
+
+```ts
+import { preprocessForSerialization } from './index'
+
+class BaseClass {
+  readonly foo = 'foo'
+  get bar() {
+    return 'bar'
+  }
+
+ printFooBar() {
+   console.log('foobar')
+ }
+}
+
+
+class SubClass extends BaseClass {
+  baz = 'baz' 
+}
+
+const mySubClass = new SubClass()
+
+const preprocessedSubClass = preprocessForSerialization(mySubClass)
+
+preprocessedSubClass.bar // is "bar"
+preprocessedSubClass.baz // is "baz"
+preprocessedSubClass.foo // is "foo"
+preprocessedSubClass.printFooBar // is undefined
+
+// only accurate if going through postMessage
+preprocessedSubClass instanceof SubClass // is false
+preprocessedSubClass instanceof BaseClass // is false
+preprocessedSubClass instanceof Object // is true
+```
+
+Arrays are also considered objects, and are processed similarly by attempting to serialize each index to a new array literal.
+
+### Complex Serialization of Logs
+
+Currently, most serialization follows the rules explained above, with a current exception being Cypress's log messages generated from `log:added` and `log:changed` events. These messages are objects that contain DOM elements, functions, and other unserializable values. We attempt to pre/post process these log messages to accurate display snapshots, as well as console props that are printed when a snapshot is pinned
+
+#### Preprocessing
+
+When preprocessing a log, snapshots are preprocessed into a string with hydrated html state for attributes and various input values, such as `checkbox` `select`, `input`, etc to make the snapshot serializable. Other DOM element like structures, such as `$el` within log `consoleProps`, are preprocessed the same way. This is similar to how [Percy serializes the DOM](https://github.com/percy/cli/blob/master/packages/dom/src/serialize-inputs.js).
+
+Additional preprocessing occurs for `consoleProps.table`, where the function output is preprocessed into a serializable object, ultimately to be wrapped back into a function when received from `postMessage`
+
+#### Postprocessing/Reification
+
+When the preprocessed log is sent through `postMessage` and received by the primary Cypress instance, the log is postprocessed. This process includes recreating the stringified snapshot as as DOM element, and inserting the snapshot, and it's styles, into the primary Cypress instance's snapshot map to be associated with its appropriate command log. This way, the correct snapshot is pinned when time travel debugging is leveraged after the test has run. `consoleProps` are recreated, with DOM elements being calculated lazily via object getters. This method works because The element(s) is only recreated once the snapshot is pinned to the page, correctly matching the element in the snapshot. This also means that if an array of elements is present, the array is actually a proxy to an object with the keys being the index number.
```

Comment:
```suggestion
When the preprocessed log is sent through `postMessage` and received by the primary Cypress instance, the log is postprocessed. This process includes recreating the stringified snapshot as a DOM element and inserting the snapshot and its styles into the primary Cypress instance's snapshot map to be associated with its appropriate command log. This way, the correct snapshot is pinned when time travel debugging is leveraged after the test has run. `consoleProps` are recreated, with DOM elements being calculated lazily via object getters. This method works because the element(s) is/are only recreated once the snapshot is pinned to the page, correctly matching the element in the snapshot. This also means that if an array of elements is present, the array is actually a proxy to an object with the keys being the index number.
```